### PR TITLE
Adds functionality to visualize hypergraphs similar to HyperNetX

### PR DESCRIPTION
**Pull request by metalcyanide**
I also added the testcases discussed in the issue #27  to playtest_utils.py

### CODE REVIEW

**Comment by matt-lourens** [comment]
@metalcyanide From a few initial tests this looks awesome! I'll check the implementation details in a bit, but this is reproducing what I wanted, great stuff :D

There is one case that we might want to change a little bit (depending on difficulty to do so). Consider:
```python
primitive = Qinit(9) + Qcycle(stride=8, step=4, offset=0, mapping=Qunitary(arity=3))
f,a = plot_motif(primitive.head)
print(primitive.head.E)
[(1, 9, 8), (5, 4, 3), (9, 8, 7)]
```
I  like what happens, it gives us some information about the edge order (the drawing path is correct, edges should cross like that). The only small issue is that the loop around the middle node is not completely visible (the one that goes from 5 around 4 to 3, the one that goes back from 3 to 5 through 4 is fine). There are also those two white gaps around node 5 and 3. Would it be hard to fix?

```python
primitive = Qinit(9) + Qcycle(stride=4, step=6, offset=3, mapping=Qunitary(arity=4))
f,a = plot_motif(primitive.head)
print(primitive.head.E)
```
This one is clearer, it just seems like the endpoints aren't meeting properly.
___
With those gaps removed and loops shown, as far as I can tell the requirements for the issue is met. I have an idea which is scope creep and not required to close the issue (so you don't have to do it). But what I thought would be cool is if the edge around the start node and end node are particular colors. This way the order of nodes in the edge will be somewhat visible. Maybe all edges can have 3 color parts, one color around the start, one during the middle loops and then the other around the final node. Do you think something like this would be difficult to implement? Like I said you don't have to implement this, but let me know what you think of the idea.

File: hierarqcal/utils.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 import numpy as np
```
Comment (matt-lourens) [inline hierarqcal/utils.py:14]: The edges are sometimes off screen, there is xlim and ylims of the plot set on lines 96. I think you can just increase the epsilon value there to ensure the edges are always visible.

Comment (metalcyanide) [reply hierarqcal/utils.py:14]: Will update this.

**Comment by metalcyanide** [comment]
> @metalcyanide From a few initial tests this looks awesome! I'll check the implementation details in a bit, but this is reproducing what I wanted, great stuff :D
> 
> There is one case that we might want to change a little bit (depending on difficulty to do so). Consider:
> 
> ```python
> primitive = Qinit(9) + Qcycle(stride=8, step=4, offset=0, mapping=Qunitary(arity=3))
> f,a = plot_motif(primitive.head)
> print(primitive.head.E)
> [(1, 9, 8), (5, 4, 3), (9, 8, 7)]
> ```
> 
> I like what happens, it gives us some information about the edge order (the drawing path is correct, edges should cross like that). The only small issue is that the loop around the middle node is not completely visible (the one that goes from 5 around 4 to 3, the one that goes back from 3 to 5 through 4 is fine). There are also those two white gaps around node 5 and 3. Would it be hard to fix?

Thanks for pointing it out. I assumed step <= arity. It happening because of the step > arity, willl make changes accordingly. 

> 
> ```python
> primitive = Qinit(9) + Qcycle(stride=4, step=6, offset=3, mapping=Qunitary(arity=4))
> f,a = plot_motif(primitive.head)
> print(primitive.head.E)
> ```
> 
> This one is clearer, it just seems like the endpoints aren't meeting properly.
> 
> With those gaps removed and loops shown, as far as I can tell the requirements for the issue is met. I have an idea which is scope creep and not required to close the issue (so you don't have to do it). But what I thought would be cool is if the edge around the start node and end node are particular colors. This way the order of nodes in the edge will be somewhat visible. Maybe all edges can have 3 color parts, one color around the start, one during the middle loops and then the other around the final node. Do you think something like this would be difficult to implement? Like I said you don't have to implement this, but let me know what you think of the idea.

Adding this functionality of coloring should easier.

**Comment by metalcyanide** [comment]
@matt-lourens Updated functionality based on your comments.

**Comment by matt-lourens** [comment]
> @matt-lourens Updated functionality based on your comments.

Hey @metalcyanide sorry I'm trying to get around to everything :D. I'll have time later today to check your new code out.

File: hierarqcal/utils.py [hunk 2]
```diff
@@ -14,6 +14,7 @@
...
 import matplotlib.pyplot as plt
```
Comment (matt-lourens) [inline hierarqcal/utils.py:15]: Remember to format the file with black

File: hierarqcal/utils.py [hunk 3]
```diff
@@ -144,10 +147,81 @@ def plot_motif(
             )
             ax.add_patch(arrow)
         else:
-            # TODO hyper graph visual (rubber band)
-            raise NotImplementedError(
-                f"Plotting {len(edge)}-ary edges are not implemented yet"
-            )
+            end_points = []
+            for i in range(len(edge)):
```
Comment (matt-lourens) [inline hierarqcal/utils.py:151]: Thanks for adding the colors! Just as I mentioned this morning, the node colors has specific meaning, so I originally meant that it's the edge that should be broken up into three color parts. We can just color those loops that gets added around the nodes to indicate, start, middle, end

File: hierarqcal/utils.py [hunk 4]
```diff
@@ -144,10 +147,81 @@ def plot_motif(
...
+                
+            for i in range(len(edge)):
+                ket_s = np.array(pos[edge[i]])  # source vector
+                ket_t = np.array(pos[edge[(i+1)%len(edge)]])  # target vector
+
+                par_vec = ket_t-ket_s
+                
+                orth_vec = np.array([par_vec[1], -par_vec[0]])*(1/np.linalg.norm(par_vec))
+                
+                if primitive.arity < primitive.step:
```
Comment (matt-lourens) [inline hierarqcal/utils.py:173]: Can you add a comment here to explain why this is done

File: hierarqcal/utils.py [hunk 5]
```diff
@@ -144,10 +147,81 @@ def plot_motif(
...
+
+                end_point_check = np.array([center[0] + t_radius*pad_ratio * np.cos(angle_ref+angle), 
+                                            center[1] + t_radius*pad_ratio * np.sin(angle_ref+angle)])
+                if np.linalg.norm(end_point_check - end) > 1e-3:
+                    angle = 2*np.pi - angle
+                arc_angles = np.linspace(angle_ref, angle_ref+angle, 20)
+                arc_xs = center[0] + t_radius*pad_ratio * np.cos(arc_angles)
+                arc_ys = center[1] + t_radius*pad_ratio * np.sin(arc_angles)
+
+                plt.plot(arc_xs, arc_ys, lw=edge_width, color='black')
```
Comment (matt-lourens) [inline hierarqcal/utils.py:221]: This is very cool!

**Review by matt-lourens — COMMENTED** [review]
Awesome! I'm happy with the code, and the tests I threw at it works. It's just changing the start middle end color indications to be parts of the edge (not nodes) and then some black formatting and a comment. Once those are in, I'll merge the PR

Nice job!

File: hierarqcal/utils.py (hunk 4 above)
Comment (metalcyanide) [reply hierarqcal/utils.py:173]: Sure will add the comment.

File: hierarqcal/utils.py (hunk 2 above)
Comment (metalcyanide) [reply hierarqcal/utils.py:15]: The default is already black, if i understand it correctly. Can you let you know exactly what do you to fix for the matplotlib style?

File: hierarqcal/utils.py (hunk 3 above)
Comment (metalcyanide) [reply hierarqcal/utils.py:151]: Added.

File: hierarqcal/utils.py (hunk 2 above)
Comment (matt-lourens) [reply hierarqcal/utils.py:15]: Sorry, I meant [black codeformatter](https://black.readthedocs.io/en/stable/), nothing to do with matplotlib, I just wanted to comment on the top of the file xD. 

Should be as simple as pip install black -> format document in vs code

Comment (metalcyanide) [reply hierarqcal/utils.py:15]: Done. Added changes after that.

**Pull request merged**